File: many_models.py
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression

df = pd.read_csv("gapminder.csv")
by_country = df.groupby(by = ['country','continent'])

def country_model(data):
    x = data.year.values.reshape(-1, 1)
    y = data.lifeExp.values.reshape(-1,1)
    model = LinearRegression().fit(x, y)
    r2 = model.score(x, y)
    return r2

(by_country
        .apply(country_model)
        .reset_index()
        .rename(columns = {0: "r2"})
        .sort_values(by = ['r2']))

File: many_models.R
#install.packages("purrr")
#install.packages("dplyr")
#install.packages("magrittr")
#install.packages("gapminder")
#install.packages("modelr")

library(dplyr)
library(magrittr)
library(purrr)
library(tidyr)
library(gapminder)
library(modelr)

gapminder %>%
  dplyr::group_by(country, continent, year) %>%
  dplyr::summarise(m = mean(lifeExp))

# group data
by_country <- gapminder %>% 
  dplyr::group_by(country, continent) %>% 
  tidyr::nest()

by_country$data[1]

# define linear model function
country_model <- function(df) {
  lm(lifeExp ~ year, data = df)
}

# key line. 
by_country <- by_country %>% 
  dplyr::mutate(model = purrr::map(data, country_model)) %>% 
  dplyr::mutate(resids = purrr:::map2(data, model, modelr::add_residuals))

# sort by largest residuals
by_country %>% 
  tidyr::unnest(resids) %>% 
  dplyr::arrange(dplyr::desc(abs(resid)))

